File: xssfinder-executor-java/src/main/java/org/xssfinder/scanner/PageDefinitionFactory.java
package org.xssfinder.scanner;

import org.xssfinder.CrawlStartPoint;
import org.xssfinder.remote.MethodDefinition;
import org.xssfinder.remote.PageDefinition;

import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class PageDefinitionFactory {
    private final MethodDefinitionFactory methodDefinitionFactory;
    private final Map<Class<?>, PageDefinition> pageDefinitionCache;

    public PageDefinitionFactory(
            MethodDefinitionFactory methodDefinitionFactory,
            Map<Class<?>, PageDefinition> pageDefinitionCache
    ) {
        this.methodDefinitionFactory = methodDefinitionFactory;
        this.pageDefinitionCache = pageDefinitionCache;
    }

    public PageDefinition createPageDefinition(
            Class<?> pageClass,
            Set<Class<?>> knownPageClasses,
            ThriftToReflectionLookup lookup
    ) {
        if (!pageDefinitionCache.containsKey(pageClass)) {
            String identifier = pageClass.getCanonicalName();
            boolean isCrawlStartPoint = pageClass.isAnnotationPresent(CrawlStartPoint.class);
            Map<MethodDefinition, Method> methodMapping = new HashMap<MethodDefinition, Method>();
            PageDefinition pageDefinition = new PageDefinition(identifier, methodMapping.keySet(), isCrawlStartPoint);
            pageDefinitionCache.put(pageClass, pageDefinition);
            lookup.putPageClass(pageDefinition.getIdentifier(), pageClass);

            // Add the methods after creating the page & putting the mapping in the cache, in case any of the methods
            // are circular (i.e. return the type they're defined on).
            methodMapping.putAll(getMethodDefinitions(pageClass, knownPageClasses, lookup));

            for (Map.Entry<MethodDefinition, Method> entry : methodMapping.entrySet()) {
                lookup.putMethod(entry.getKey().getIdentifier(), entry.getValue());
            }
        }
        return pageDefinitionCache.get(pageClass);
    }

    private Map<MethodDefinition, Method> getMethodDefinitions(
            Class<?> pageClass,
            Set<Class<?>> knownPageClasses,
            ThriftToReflectionLookup lookup
    ) {
        Map<MethodDefinition, Method> methods = new HashMap<MethodDefinition, Method>();
        for (Method method : pageClass.getDeclaredMethods()) {
            if (knownPageClasses.contains(method.getReturnType())) {
                MethodDefinition methodDefinition = methodDefinitionFactory.createMethodDefinition(
                        method,
                        this,
                        knownPageClasses,
                        lookup
                );
                methods.put(methodDefinition, method);
            }
        }
        return methods;
    }
}

File: xssfinder-php-executor/src/XssFinder/Scanner/PageDefinitionFactory.php
<?php

namespace XssFinder\Scanner;

use XssFinder\Annotations\Annotations;
use XssFinder\PageDefinition;

class PageDefinitionFactory
{
    /** @var \XssFinder\Scanner\MethodDefinitionFactory */
    private $_methodDefinitionFactory;
    /** @var ReflectionHelper */
    private $_reflectionHelper;
    /** @var array */
    private $_knownPageClassNames;

    /**
     * @param MethodDefinitionFactory $methodDefinitionFactory
     * @param ReflectionHelper $reflectionHelper
     * @param array $knownPageClassNames
     */
    public function __construct(
        MethodDefinitionFactory $methodDefinitionFactory,
        ReflectionHelper $reflectionHelper,
        array $knownPageClassNames
    ) {
        $this->_methodDefinitionFactory = $methodDefinitionFactory;
        $this->_reflectionHelper = $reflectionHelper;
        $this->_knownPageClassNames = $knownPageClassNames;
    }

    /**
     * @param $className string Fully qualified name of the page object class
     * @param ThriftToReflectionLookup $lookup
     * @return \XssFinder\PageDefinition
     */
    public function createPageDefinition($className, ThriftToReflectionLookup $lookup)
    {
        $reflectionClass = new \ReflectionClass($className);

        $pageDefinition = new PageDefinition();
        $pageDefinition->identifier = '\\' . $reflectionClass->getName();
        $pageDefinition->methods = $this->_getMethodDefinitions($reflectionClass, $lookup);
        $pageDefinition->crawlStartPoint = $this->_isCrawlStartPoint($reflectionClass);

        $lookup->putPageClass($pageDefinition->identifier, $reflectionClass);

        return $pageDefinition;
    }

    /**
     * @param $reflectionClass \ReflectionClass
     * @param ThriftToReflectionLookup $lookup
     * @return array
     */
    private function _getMethodDefinitions(\ReflectionClass $reflectionClass, ThriftToReflectionLookup $lookup)
    {
        $methodDefinitions = array();
        foreach ($reflectionClass->getMethods() as $method) {
            $returnClassName = $this->_reflectionHelper->getReturnType($method);
            if ($returnClassName == null) {
                continue;
            }
            $methodDefinition =
                $this->_createMethodDefinitionIfReturnClassIsPage($returnClassName, $method);
            if ($methodDefinition !== null) {
                $methodDefinitions[] = $methodDefinition;
                $lookup->putMethod($methodDefinition->identifier, $method);
            }
        }
        return $methodDefinitions;
    }

    /**
     * @param $returnClassName string
     * @param $method \ReflectionMethod
     * @return null|\XssFinder\MethodDefinition
     */
    private function _createMethodDefinitionIfReturnClassIsPage($returnClassName, \ReflectionMethod $method)
    {
        if (in_array($returnClassName, $this->_knownPageClassNames)) {
            return $this->_methodDefinitionFactory->createMethodDefinition($method);
        }
        return null;
    }

    private function _isCrawlStartPoint(\ReflectionClass $reflectionClass)
    {
        $annotationsManager = Annotations::getConfiguredManager();
        $annotations = $annotationsManager->getClassAnnotations($reflectionClass, '@crawlStartPoint');
        return !empty($annotations);
    }
}